packages/socket-mode/src/SocketModeClient.ts

```diff
@@ -694,8 +694,12 @@ export class SocketModeClient extends EventEmitter {
    * `onmessage` handler for the client's WebSocket.
    * This will parse the payload and dispatch the relevant events for each incoming message.
    */
-  protected async onWebSocketMessage({ data }: { data: string }): Promise<void> {
-    this.logger.debug(`Received a message on the WebSocket: ${data}`);
+  protected async onWebSocketMessage(data: WebSocket.RawData, isBinary: boolean): Promise<void> {
```

Comment:
Is it safe enough to assume that this data cannot be undefined | null? If we're unsure about it, we may want to check the value existence in the beginning. of this internal method.

Reply:
Maybe it is worth adding some tests to answer this question? The coverage numbers for this package are poor (30%). To be honest I'm nervous making this web socket dependency major version upgrade without a basic integration test with a websocket peer.

I did test these changes manually with a real app and a few events were handled fine, but having more tests would give us more confidence to make changes in this project.

Reply:
Adding tests is great! but just making the code robust enough to deal with any patterns (meaning RawData | undefined | null) would sound enough for me.

Reply:
@seratch I added a couple of basic integration tests, if you could please have a look.
I think those uncovered a bug, or at least a documentation discrepancy... I had to introduce an additional sleep or wait for the `connected` event to be emitted to be able to `start()` and then `disconnect()` the socket mode client gracefully (see the test code for details). If I omit this and just call:

```
await client.start();
await client.disconnect();
```

.. I get the following error and output:

```
➜ npm run test:integration

> @slack/socket-mode@1.3.3 test:integration
> mocha --config .mocharc.json --timeout 10000 test/integration.spec.js



  Integration tests with a WebSocket server
    ✔ connects to a server
    1) connects to a server


  1 passing (35ms)
  1 failing

  1) Integration tests with a WebSocket server
       connects to a server:
     Uncaught Error: WebSocket was closed before the connection was established
```

I also think the README is slightly off with respect to the events that are emitted. I tried listening to the `ready` event, which we [document here](https://github.com/slackapi/node-slack-sdk/blob/main/packages/socket-mode/README.md#lifecycle-events) instead of the `connected` event, but it never gets raised. I believe this is because [the only events that are emitted are the ones on the "main" state machine in this project](https://github.com/slackapi/node-slack-sdk/blob/main/packages/socket-mode/src/SocketModeClient.ts#L249-L259), whereas [the `ready` event is part of the "connected" sub-machine](https://github.com/slackapi/node-slack-sdk/blob/main/packages/socket-mode/src/SocketModeClient.ts#L240) (and state transitions in this sub-machine are not emitted).

Let me know if you have thoughts on that: we could either remove the `ready` event from the docs, or we can emit events from the sub-machine(s) as well.

Since this PR is part of upcoming major release, we can introduce backwards incompatible changes at this time.

Reply:
@seratch one more thing: I added a `describe.skip` block of a couple of failure tests: sending a `null` WS message and sending an empty string WS message.

The null message gets treated as binary, is logged out as an ERROR, and is essentially ignored in the implementation in this PR.

The empty string message raises a JSON parsing exception (as is expected), which is logged as an ERROR, but is not catastrophic to the client (the client continues to function).

If you have thoughts on these, let me know. If you don't think they are useful, we can remove them.

Reply:
@filmaj Thanks for the tests! 

>The ready state topic

I remember that when I improved this module's internals a lot to make it stable, I had to move the ready state to the "connected" sub state machine. Thus, I agree that the documents are now outdated. Could you remove the "ready" state from the docs? I am confident that no one has been relying on it (because it never worked) but we can quickly mention it in the migration guide as necessary.

> the null / empty string tests

Thanks for verifying this! The only slight fear I still have is that the possibility to receive invalid text message from Slack servers due to its incidents/outage/misbehaviors (since the WS server is not a simple hosted OSS code and it has its own custom logic such as disconnection warnings etc., there is a certain risk to cause unexpected data delivery). Writing lots of error logs of JSON parsing error could be a false monitoring issue on the customer side (they may be PDed due to that). Thus, I think ignoring these invalid partterns without raising errors coluld be beneficial for SDK users. But if it takes your time a lot, we can do this later on. It's not a breaking change but an internal improvements. We can do it at any time!

Reply:
> Writing lots of error logs of JSON parsing error could be a false monitoring issue on the customer side (they may be PDed due to that). Thus, I think ignoring these invalid partterns without raising errors coluld be beneficial for SDK users.

Thanks for your input! In this case I will set any logs for these results as DEBUG level, and see if I can modify the tests to verify this behaviour.

I will also add a few basic tests to verify the various event names/states we document are being emitted, and remove the `ready` one.